### PR TITLE
Regenerating Volfiles on Volume Start and Stop

### PR DESCRIPTION
**Pull request by aravindavk**
Cluster level volfiles should not contain the information about Stopped
Volumes. Added checks for the same while generating Cluster level
volfiles. But when Volume starts, these volfiles will not contain the
started Volume's info. iTo address this issue added Volfiles re-generation
after Volume Stop and Start.

Closes: #1005
Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

**Comment by aravindavk** [comment]
Create two Volumes

    # glustercli  volume create --name gv1 --replica 2 \
        gluster1.example.com:/bricks/b1 \
        gluster1.example.com:/bricks/b2 \
        --force
    # glustercli  volume create --name gv2 --replica 2 \
        gluster1.example.com:/bricks/b3 \
        gluster1.example.com:/bricks/b4 \
        --force

Before starting the Volumes

    # curl http://localhost:24007/v1/volfiles/gluster/glustershd

    volume glustershd
        type debug/io-stats
    end-volume

After starting one Volume,

    # glustercli volume start gv1
    # curl http://localhost:24007/v1/volfiles/gluster/glustershd

    volume gv1-replicate-0-client-1
        type protocol/client
    end-volume

    volume gv1-replicate-0
        type cluster/replicate
        subvolumes gv1-replicate-0-client-0 gv1-replicate-0-client-1
    end-volume

    volume glustershd
        type debug/io-stats
        subvolumes gv1-replicate-0
    end-volume


After starting two Volumes

    # glustercli volume start gv2
    # curl http://localhost:24007/v1/volfiles/gluster/glustershd

    volume gv1-replicate-0-client-0
        type protocol/client
    end-volume
     
    volume gv1-replicate-0-client-1
        type protocol/client
    end-volume
     
    volume gv1-replicate-0
        type cluster/replicate
        subvolumes gv1-replicate-0-client-0 gv1-replicate-0-client-1
    end-volume
     
    volume gv2-replicate-0-client-0
        type protocol/client
    end-volume
     
    volume gv2-replicate-0-client-1
        type protocol/client
    end-volume
     
    volume gv2-replicate-0
        type cluster/replicate
        subvolumes gv2-replicate-0-client-0 gv2-replicate-0-client-1
    end-volume
     
    volume glustershd
        type debug/io-stats
        subvolumes gv1-replicate-0 gv2-replicate-0
    end-volume


After stopping one Volume,

    # glustercli volume stop gv1
    # curl http://localhost:24007/v1/volfiles/gluster/glustershd

    volume gv2-replicate-0-client-0
        type protocol/client
    end-volume

    volume gv2-replicate-0-client-1
        type protocol/client
    end-volume

    volume gv2-replicate-0
        type cluster/replicate
        subvolumes gv2-replicate-0-client-0 gv2-replicate-0-client-1
    end-volume

    volume glustershd
        type debug/io-stats
        subvolumes gv2-replicate-0
    end-volume

File: glusterd2/commands/volumes/volume-start.go [hunk 1]
```diff
@@ -104,22 +107,28 @@ func volumeStartHandler(w http.ResponseWriter, r *http.Request) {
...
+		{
+			DoFunc:   "vol-start.UpdateVolinfo",
+			UndoFunc: "vol-start.UpdateVolinfo.Undo",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+		},
 	}
 
-	if err := txn.Ctx.Set("volinfo", volinfo); err != nil {
+	if err := txn.Ctx.Set("oldvolinfo", volinfo); err != nil {
 		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
```
Comment (rishubhjain) [inline glusterd2/commands/volumes/volume-start.go]: May be we can add logging here.

File: glusterd2/commands/volumes/volume-start.go [hunk 2]
```diff
@@ -104,22 +107,28 @@ func volumeStartHandler(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 
-	if err := txn.Do(); err != nil {
-		logger.WithError(err).WithField(
-			"volume", volname).Error("transaction to start volume failed")
+	volinfo.State = volume.VolStarted
+
+	if err := txn.Ctx.Set("volinfo", volinfo); err != nil {
 		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
```
Comment (rishubhjain) [inline glusterd2/commands/volumes/volume-start.go]: Logging.

File: glusterd2/commands/volumes/volume-start.go (hunk 1 above)
Comment (prashanthpai) [reply glusterd2/commands/volumes/volume-start.go]: It's already logged internally at `Error` log level.

File: glusterd2/commands/volumes/volume-start.go (hunk 2 above)
Comment (prashanthpai) [reply glusterd2/commands/volumes/volume-start.go]: It's logged internally at `Error` log level.

File: glusterd2/commands/volumes/volume-start.go (hunk 1 above)
Comment (prashanthpai) [reply glusterd2/commands/volumes/volume-start.go]: See https://github.com/gluster/glusterd2/blob/v4.1.0/glusterd2/transaction/context.go#L69

**Pull request merged**